### PR TITLE
Introduce wrapper library

### PR DESCRIPTION
**Pull request by bartekxk**
#900

### CODE REVIEW

File: docs/index.rst [hunk 1]
```diff
@@ -34,6 +34,7 @@ Current CK library are structured into 4 layers:
 * "Templated Tile Operators" layer
 * "Templated Kernel and Invoker" layer
 * "Instantiated Kernel and Invoker" layer
+* "Wrapper for tensor tranforms operations"
```
Comment (bwroblew) [inline docs/index.rst]:
```suggestion
* "Wrapper for tensor transform operations"
```

File: docs/wrapper.rst [hunk 2]
```diff
@@ -0,0 +1,52 @@
+===============
+Wrapper
+===============
+
+-------------------------------------
+Description
+-------------------------------------
+Note: Wrapper is currently under development. At the moment, its functionality 
+is limited.
+
```
Comment (bwroblew) [inline docs/wrapper.rst]:
```suggestion

.. note::

    Wrapper is currently under development. At the moment, its functionality is limited.

```

File: docs/wrapper.rst [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            std::cout << layout(ck::make_tuple(h, w)) << " ";
+        }
+        std::cout << std::endl;
+    }
+
+Output::
+
+    dims:4,(2,4) strides:2,(1,8)
+    Print2d
```
Comment (bwroblew) [inline docs/wrapper.rst]: Where does this `Print2d` come from?

File: include/ck/wrapper/layout.hpp [hunk 4]
```diff
@@ -224,36 +205,32 @@ struct Layout
...
+        const auto unrolled_shape   = UnrollNestedTuple(shape);
+        const auto unrolled_strides = UnrollNestedTuple(strides);
+        static_assert(unrolled_shape.Size() == unrolled_strides.Size(),
+                      "Size of strides and shape are not consistent.");
+        return make_naive_tensor_descriptor(unrolled_shape, unrolled_strides);
     }
 
     public:
-    using NaiveDescriptorType = remove_cvref_t<decltype(MakeNaiveDescriptor(Shape{}, Strides{}))>;
+    // If stride not passed, deduce from GenerateColumnMajorPackedStrides
```
Comment (bwroblew) [inline include/ck/wrapper/layout.hpp]:
```suggestion
    // If stride not is passed, deduce it from `GenerateColumnMajorPackedStrides`.
```

Comment (LisaDelaney) [reply include/ck/wrapper/layout.hpp]: If the stride is not passed, you can infer it from `GenerateColumnMajorPackedStrides`.

File: docs/wrapper.rst (hunk 2 above)
Comment (LisaDelaney) [reply docs/wrapper.rst]: The wrapper is under development and its functionality is limited.

File: include/ck/wrapper/layout.hpp [hunk 5]
```diff
@@ -32,21 +18,19 @@ namespace tensor_transform_wrapper {
...
     static constexpr auto I1 = Number<1>{};
 
-    template <typename T>
-    using is_tuple = decltype(std::declval<T&>().IsTuple());
-
     // Generate packed (column-major) strides if not passed
     template <typename... Ts>
     __host__ __device__ constexpr static auto
-    GenerateColumnMajorPackedStrides(const Tuple<Ts...>& tuple)
+    GenerateColumnMajorPackedStrides(const Tuple<Ts...>& shape)
```
Comment (aosewski) [inline include/ck/wrapper/layout.hpp:31]: Please add documentation. Please describe what kind of input arguments are supported. Give example.

File: include/ck/wrapper/layout.hpp [hunk 6]
```diff
@@ -224,36 +205,32 @@ struct Layout
             static_assert(Tuple<ShapeDims...>::Size() == Tuple<IdxDims...>::Size(),
                           "Idx rank and Shape rank must be the same (except 1d).");
             // Unroll while IdxDims is nested
-            const auto unrolled_shape_via_idx = UnrollShapeViaIdx(shape, idx);
+            const auto unrolled_shape_via_idx = AlignShapeToIdx(shape, idx);
```
Comment (aosewski) [inline include/ck/wrapper/layout.hpp]:
```suggestion
            const auto aligned_shape = AlignShapeToIdx(shape, idx);
```

File: include/ck/wrapper/layout.hpp [hunk 7]
```diff
@@ -169,8 +150,8 @@ struct Layout
     // Example shape:    2, (2, 2), 2, (2, 2)
     // Merged shape:     2,      4, 2,      4
     template <typename... ShapeDims, typename... IdxDims, typename DescriptorToMerge>
-    __host__ __device__ constexpr static auto
-    MakeMerges(const Tuple<ShapeDims...>& shape, const Tuple<IdxDims...>&, DescriptorToMerge& desc)
+    __host__ __device__ constexpr static auto CreateMergedDescriptor(
```
Comment (aosewski) [inline include/ck/wrapper/layout.hpp:153]: Please add documentation. Describe how you create a new descriptor. ie: which dimensions you merge from original descriptor and when you're merging them.

File: test/wrapper/test_layout.cpp [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+    EXPECT_EQ(ck::wrapper::size(layout_runtime), d4 * d3 * d2 * d1 * d0);
+    EXPECT_EQ(ck::wrapper::size(layout_compiletime), d4 * d3 * d2 * d1 * d0);
+
+    // Size of dims
+    EXPECT_EQ(ck::wrapper::size<0>(layout_runtime), d4 * d3 * d2);
+    EXPECT_EQ(ck::wrapper::size<0>(layout_compiletime), d4 * d3 * d2);
+    EXPECT_EQ(ck::wrapper::size<1>(layout_runtime), d1 * d0);
+    EXPECT_EQ(ck::wrapper::size<1>(layout_compiletime), d1 * d0);
+
+    // Acces via new layout (using get on layout)
```
Comment (aosewski) [inline test/wrapper/test_layout.cpp]:
```suggestion
    // Access through new layout (using get with layout object)
```

File: include/ck/wrapper/layout.hpp (hunk 7 above)
Comment (bartekxk) [reply include/ck/wrapper/layout.hpp:153]: This is private method so I just added a comments.

File: include/ck/wrapper/layout.hpp (hunk 5 above)
Comment (bartekxk) [reply include/ck/wrapper/layout.hpp:31]: This is private method so I just added a comments.

File: docs/wrapper.rst (hunk 3 above)
Comment (bartekxk) [reply docs/wrapper.rst]: REmoved.

**Review by LisaDelaney — APPROVED** [review]

File: include/ck/wrapper/layout.hpp [hunk 9]
```diff
@@ -144,7 +144,7 @@ struct Layout
             desc, make_tuple(make_merge_transform(merge_elems)), lower_dims, upper_dims);
     }
 
-    // Merge nested shape dims
+    // Merge nested shape dims. Merge nested shape dims when idx is also nested.
```
Comment (aosewski) [inline include/ck/wrapper/layout.hpp:147]:
```suggestion
    // Merge nested shape dims when corresponding index is also nested.
```

**Review by aosewski — APPROVED** [review]

File: include/ck/wrapper/layout.hpp (hunk 9 above)
Comment (bartekxk) [reply include/ck/wrapper/layout.hpp:147]: Will be fixed in next PR, thanks

**Pull request merged**